src/source/source_state.js

```diff
@@ -16,50 +16,137 @@ export type LayerFeatureStates = {[layer: string]: FeatureStates};
 class SourceFeatureState {
     state: LayerFeatureStates;
     stateChanges: LayerFeatureStates;
+    deletedStates: {};
```

Comment:
The goal of the merging/reconciling logic in this class isn't immediately clear, I think an introductory comment at the top could help make the rest a lot easier to read. Something along the lines of:

- we can’t just change the state directly because we want to batch changes to the tiles
- we store all updates in a single map, doing an overwrite/merge with previous updates
- we store deletes in a separate map that we use to “mask out” any changes in the update map that have since been removed
- whenever we update something, we make sure to take it out of the deleted map if it’s there.